OpenRA.Mods.Common/Graphics/ContrailRenderable.cs

```diff
@@ -73,11 +73,28 @@ public void Render(WorldRenderer wr)
 			// Start of the first line segment is the tail of the list - don't smooth it.
 			var curPos = trail[Index(next - skip - 1)];
 			var curColor = color;
-			for (var i = 0; i < length - skip - 4; i++)
+
+			for (var i = 1; i < renderLength; i++)
 			{
-				var j = next - skip - i - 2;
-				var nextPos = Average(trail[Index(j)], trail[Index(j - 1)], trail[Index(j - 2)], trail[Index(j - 3)]);
-				var nextColor = Exts.ColorLerp(i * 1f / (length - 4), color, Color.Transparent);
+				var j = next - skip - 1 - i;
+				WPos nextPos;
+				int maxSmoothLength = 4;
+
+				var alpha = ((renderLength - i) * color.A + renderLength - 1) / renderLength;
+				var nextColor = Color.FromArgb(alpha, color);
+
+				var nextX = 0L;
+				var nextY = 0L;
+				var nextZ = 0L;
+				var k = 0;
+				for (; k < renderLength - i && k < maxSmoothLength; k++)
+				{
+					nextX += trail[Index(j - k)].X;
```

Comment:
Calculate Index once. Access array once.

`var index = Index(j- k);` or rather `trail[Index(j - k) = new WPos(y, x, z)`.

Reply:
Last year I considered making trail a rotating buffer. 

Seemed like a shame to calculate all elements of the trail array each render frame - just to add one additional item. 

A more simple straigtforward alternative could be to precaculate the colors once. Since you will always will want to display the full trail length. So you might as well calculate the expensive colors up front add them a lookup array[i].

Reply:
It already is a rotating buffer?

Reply:
It is actually a rotation buffer already. And will fix on accessing.

Reply:
Yeah i read `Index()` after. I guess i back then looked at not again calculating curPos and nextPos for the start of the trail - since that would each time be the same. Only the head falls off - and the average over the tail three could change?

Reply:
I change to if don't have enough position to smooth, like x < 4, then use x position to smooth to cancel the solid delay of high speed projectile.

Reply:
Oh and I split up x, y, z here is for readability. Do you have better idea?

Reply:
I cache the `trail[Index(j - k)]` now